java-frontend/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -725,27 +725,48 @@ private void buildLabeledStatement(LabeledStatementTree labeledStatement) {
   }
 
   private void buildSwitchStatement(SwitchStatementTree switchStatementTree) {
+    buildSwitchExpression(switchStatementTree.asSwitchExpression(), switchStatementTree);
+  }
+
+  private void buildSwitchExpression(SwitchExpressionTree switchExpressionTree) {
+    buildSwitchExpression(switchExpressionTree, switchExpressionTree);
+  }
+
+  private void buildSwitchExpression(SwitchExpressionTree switchExpressionTree, Tree terminator) {
+    if (terminator.is(Tree.Kind.SWITCH_EXPRESSION)) {
+      // force a switch expression in the current block
+      currentBlock.elements.add(terminator);
+    }
     Block switchSuccessor = currentBlock;
     // process condition
     currentBlock = createBlock();
-    currentBlock.terminator = switchStatementTree;
+    currentBlock.terminator = terminator;
     switches.addLast(currentBlock);
-    build(switchStatementTree.expression());
+    build(switchExpressionTree.expression());
     Block conditionBlock = currentBlock;
     // process body
     currentBlock = createBlock(switchSuccessor);
     breakTargets.addLast(switchSuccessor);
     boolean hasDefaultCase = false;
-    if (!switchStatementTree.cases().isEmpty()) {
-      CaseGroupTree firstCase = switchStatementTree.cases().get(0);
-      for (CaseGroupTree caseGroupTree : Lists.reverse(switchStatementTree.cases())) {
+    if (!switchExpressionTree.cases().isEmpty()) {
+      boolean withoutFallTrough = switchWithoutFallThrough(switchExpressionTree);
+      CaseGroupTree firstCase = switchExpressionTree.cases().get(0);
+      for (CaseGroupTree caseGroupTree : Lists.reverse(switchExpressionTree.cases())) {
+        if (withoutFallTrough) {
```

Comment:
I think we could move this `if` to the place where the successor is being set (line #772)

Reply:
Nop, it does not work, depending of what is happening when building the body of the case, it result in different orders.

Reply:
ok, the code is bit messy, but it would require bigger restructuring to clean-up.